From 803f4a0b2f246df27601070040e05642a37531bd Mon Sep 17 00:00:00 2001
From: Franck Bui-Huu <fbuihuu@gmail.com>
Date: Thu, 11 Oct 2012 14:11:52 +0000
Subject: [PATCH 3/3] Make list_desktop() use the new desktop helpers

Signed-off-by: Franck Bui-Huu <fbuihuu@gmail.com>
---
 main.c |   60 +++++++++++-------------------------------------------------
 1 file changed, 11 insertions(+), 49 deletions(-)

diff --git a/main.c b/main.c
index 9514ab1..826f86b 100644
--- a/main.c
+++ b/main.c
@@ -1117,9 +1117,8 @@ static int action_window_str (Display *disp, char mode) {/*{{{*/
 }/*}}}*/
 
 static int list_desktops (Display *disp) {/*{{{*/
-    unsigned long *num_desktops = NULL;
+    int _num_desktops, *num_desktops = &_num_desktops;
     unsigned long *cur_desktop = NULL;
-    unsigned long desktop_list_size = 0;
     unsigned long *desktop_geometry = NULL;
     unsigned long desktop_geometry_size = 0;
     gchar **desktop_geometry_str = NULL;
@@ -1131,23 +1130,14 @@ static int list_desktops (Display *disp) {/*{{{*/
     gchar **desktop_workarea_str = NULL;
     gchar *list = NULL;
     int i;
-    int id;
     Window root = DefaultRootWindow(disp);
     int ret = EXIT_FAILURE;
     gchar **names = NULL;
-    gboolean names_are_utf8 = TRUE;
-    
-    if (! (num_desktops = (unsigned long *)get_property(disp, root,
-            XA_CARDINAL, "_NET_NUMBER_OF_DESKTOPS", NULL))) {
-        if (! (num_desktops = (unsigned long *)get_property(disp, root,
-                XA_CARDINAL, "_WIN_WORKSPACE_COUNT", NULL))) {
-            fputs("Cannot get number of desktops properties. "
-                  "(_NET_NUMBER_OF_DESKTOPS or _WIN_WORKSPACE_COUNT)"
-                  "\n", stderr);
-            goto cleanup;
-        }
-    }
-    
+
+    *num_desktops = get_desktop_number(disp);
+    if (num_desktops < 0)
+	    goto cleanup;
+
     if (! (cur_desktop = (unsigned long *)get_property(disp, root,
             XA_CARDINAL, "_NET_CURRENT_DESKTOP", NULL))) {
         if (! (cur_desktop = (unsigned long *)get_property(disp, root,
@@ -1159,21 +1149,9 @@ static int list_desktops (Display *disp) {/*{{{*/
         }
     }
 
-    if (options.wa_desktop_titles_invalid_utf8 || 
-            (list = get_property(disp, root, 
-            XInternAtom(disp, "UTF8_STRING", False), 
-            "_NET_DESKTOP_NAMES", &desktop_list_size)) == NULL) {
-        names_are_utf8 = FALSE;
-        if ((list = get_property(disp, root, 
-            XA_STRING, 
-            "_WIN_WORKSPACE_NAMES", &desktop_list_size)) == NULL) {
-            p_verbose("Cannot get desktop names properties. "
-                  "(_NET_DESKTOP_NAMES or _WIN_WORKSPACE_NAMES)"
-                  "\n");
-            /* ignore the error - list the desktops without names */
-        }
-    }
- 
+    /* failure is not fatal */
+    names = get_desktop_names(disp);
+
     /* common size of all desktops */
     if (! (desktop_geometry = (unsigned long *)get_property(disp, DefaultRootWindow(disp),
                     XA_CARDINAL, "_NET_DESKTOP_GEOMETRY", &desktop_geometry_size))) {
@@ -1194,21 +1172,6 @@ static int list_desktops (Display *disp) {/*{{{*/
             p_verbose("Cannot get _NET_WORKAREA property.\n");
         }
     }
-     
-    /* prepare the array of desktop names */
-    names = g_malloc0(*num_desktops * sizeof(char *));
-    if (list) {
-        id = 0;
-        names[id++] = list;
-        for (i = 0; i < desktop_list_size; i++) {
-            if (list[i] == '\0') {
-                if (id >= *num_desktops) {
-                    break;
-                }
-                names[id++] = list + i + 1;
-            }
-        }
-    }
 
     /* prepare desktop geometry strings */
     desktop_geometry_str = g_malloc0((*num_desktops + 1) * sizeof(char *));
@@ -1315,7 +1278,7 @@ static int list_desktops (Display *disp) {/*{{{*/
  
     /* print the list */
     for (i = 0; i < *num_desktops; i++) {
-        gchar *out = get_output_str(names[i], names_are_utf8);
+        gchar *out = names ? names[i] : NULL;
         printf("%-2d %c DG: %-*s  VP: %-*s  WA: %-*s  %s\n", i, i == *cur_desktop ? '*' : '-',
                 longest_str(desktop_geometry_str), desktop_geometry_str[i], 
                 longest_str(desktop_viewport_str), desktop_viewport_str[i], 
@@ -1324,7 +1287,7 @@ static int list_desktops (Display *disp) {/*{{{*/
         g_free(out);
     }
     
-    p_verbose("Total number of desktops: %lu\n", *num_desktops);
+    p_verbose("Total number of desktops: %d\n", *num_desktops);
     p_verbose("Current desktop ID (counted from zero): %lu\n", *cur_desktop);
     
     ret = EXIT_SUCCESS;
@@ -1332,7 +1295,6 @@ static int list_desktops (Display *disp) {/*{{{*/
     
 cleanup:
     g_free(names);
-    g_free(num_desktops);
     g_free(cur_desktop);
     g_free(desktop_geometry);
     g_strfreev(desktop_geometry_str);
-- 
1.7.9.2

